service/storage_proxy.cc

```diff
@@ -3725,6 +3728,25 @@ class abstract_read_executor : public enable_shared_from_this<abstract_read_exec
     lw_shared_ptr<column_family>& get_cf() {
         return _cf;
     }
+
+    // Maximum latency of a successful request made to a replica (over all requests that finished up to this point),
+    // measured using a low resolution clock (10 ms granularity).
+    // Example usage: gathering latency statistics for deciding on invoking speculative retries.
+    std::optional<clock_type::duration> max_request_latency() const {
+        return _max_request_latency;
+    }
+
+private:
+    // Save a successful replica request latency measurement.
+    void register_request_latency(clock_type::duration d) {
+        if (_max_request_latency) {
+            _max_request_latency = std::max(*_max_request_latency, d);
+        } else {
+            _max_request_latency = d;
+        }
```

Comment:
Wouldn't it be simpler and more preformant to init _max_request_latency to 0 and then unconditionally call `_max_request_latency = std::max(_max_request_latency, d);` ?

Reply:
That would of course mean _max_request_latency would no longer be an optional but just a clock_type::duration and check for whether it is present would be comparison to 0.

Reply:
@haaawk Won't the clock granularity sometimes round the latency down to zero? If so, then this would cause a change in behavior (we can't differentiate 0 latency from no latency recorded).

Can durations be negative? If so, then maybe we can use negative duration to indicate no measurement. `std::max`-ing with a negative value would work as intended.

If durations can't be negative, maybe we can introduce a bool flag which tells that the latency was recorded?

Reply:
bool flag or negative initial value would both work and both would remove the if.

Reply:
assuming duration can have negative value

Reply:
Agree about trying to eliminate the optional. Bonus if the algorithm (when consuming it) can be made to work without checking the special value.